### PR TITLE
Dameonize the dync server

### PR DESCRIPTION
**Pull request by sven1103**
The server can now be started as daemon (only, but this can be extended quickly if normal mode is necessary). The location of the PID is determined by the server's config script, and so is the umask setting  (which is 0077 atm).

### CODE REVIEW

File: dync/server.py [hunk 1]
```diff
@@ -346,5 +331,43 @@ def main():
...
+
+
+def main():
+
+    try:  # Try to load the config file
+        config = load_config(SERVER_CONFIG)
+    except FileNotFoundError:
+        log.error("Could not load configuration file {}".format(SERVER_CONFIG))
+        sys.exit(1)
+    except yaml.YAMLError as exc:
```
Comment (aseyboldt) [inline dync/server.py]: `except (yaml.YAMLError, ConfigException) as exc` or in this case probably `except Exception:` and then `log.exception("Error while reading config file: ")`

File: dync/server.py [hunk 2]
```diff
@@ -346,5 +331,43 @@ def main():
             log.info("Server stopped.")
 
 
+def print_help_msg():
+    print("usage: {} start|stop|restart".
```
Comment (aseyboldt) [inline dync/server.py]: This should go to stderr

File: dync/server.py [hunk 3]
```diff
@@ -346,5 +331,43 @@ def main():
...
+
+    if len(sys.argv) == 2:
+        if sys.argv[1] == "start":
+            dync_dameon.start(init, config)
+        elif sys.argv[1] == "stop":
+            dync_dameon.stop()
+        elif sys.argv[1] == "restart":
+            dync_dameon.restart(init, config)
+        else:
+            print("unknown command"
```
Comment (aseyboldt) [inline dync/server.py]: stderr

File: dync/server.py [hunk 4]
```diff
@@ -326,8 +313,6 @@ def main():
         log.critical("Failed to load keys", exc_info=True)
         return 1
```
Comment (aseyboldt) [inline dync/server.py]: This `return 1` is never used.

File: dync/daemon.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python
```
Comment (sven1103) [inline dync/daemon.py]: done

File: dync/server.py (hunk 4 above)
Comment (sven1103) [inline dync/server.py]: maybe you tell me what u wanted to do with `return 1`? :)

File: dync/daemon.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+        if not pid:
+            message = "pidfile {0} does not exist. " + \
+                      "Daemon not running?\n"
+            sys.stderr.write(message.format(self._pidfile))
+            return  # not an error in a restart
+
+        # Try killing the daemon process
+        try:
+            while 1:
```
Comment (aseyboldt) [inline dync/daemon.py]: Just to clarify: I think having this code in here doen't hurt. Just change the loop to something reasonabe. Killing processes is always tricky and somewhat ugly, as pids can be reused. You can check if a process is running with `os.kill(pid, 0)` if it doesn't raise an exception, that this process is alive.

Comment (aseyboldt) [inline dync/daemon.py]: The init script ties into the init system ;-)
It is responsible for starting and stopping processes, so you can make sure the deamon is started when the server starts up and cleanly shuts down if the server shuts down. Also, you can use the same tools (systemctl, service, ...) to start/stop all services on the server.

Comment (aseyboldt) [inline dync/daemon.py]: Of course this script can just call `dync-server start/stop`. See `https://github.com/qbicsoftware/dropboxhandler/blob/master/service/sysV.sh` for an (not so well written) example.

File: dync/server.py (hunk 3 above)
Comment (sven1103) [inline dync/server.py]: done

File: dync/server.py (hunk 2 above)
Comment (sven1103) [inline dync/server.py]: done

File: dync/server.py (hunk 4 above)
Comment (aseyboldt) [inline dync/server.py]: Oh, did I write this? ;-)
I guess I meant sys.exit(1). Maybe it was called as `retcode = init(...)`? Or maye this was just a bug in dropboxhandler.

Comment (aseyboldt) [inline dync/server.py]: And now that I think about it, it would probably be nicer to raise an exception and then catch that later and exit accordingly.

Comment (sven1103) [inline dync/server.py]: what do you mean with 'catch that later'? I would log the exception and sys.exit(1), as there will be no need for continuing without proper auth, right?

**Comment by aseyboldt** [comment]
Oh, and something else: For debugging it is usually much nicer to have dync-server running in a non-daemon mode. So maybe add a '-d' flag, to indicate that dync-server should start as a daemon. This will probably save quit a bit of time as soon as something goes wrong.

**Comment by aseyboldt** [comment]
@sven1103 Reviewing code is nicer than I thought. It's kind of fun to tell other people how to improve stuff, if you do not need to do it yourself. ;-)

**Comment by sven1103** [comment]
@aseyboldt  Always remember, the other side has not so much fun

**Comment by sven1103** [comment]
Honestly: Code review is so important, what I learnt these days is so valuable (jeah, receiving a comment always means more work, but mostly for the good ;))

**Comment by aseyboldt** [comment]
Well, I guess i do have some sardistic inclinations after all. ;-)

File: dync/server.py (hunk 4 above)
Comment (sven1103) [inline dync/server.py]: Done

File: dync/server.py [hunk 7]
```diff
@@ -301,17 +302,9 @@ def _check_config(config):
...
-        log.error("Could not load logger settings because of: {}".format(e))
-        sys.exit(1)
+    logging.config.dictConfig(config['logging'])
 
-    try:
-        auth, server_keys = prepare_auth(ctx, os.path.expanduser('~/.dync'))
-    except Exception:
-        log.critical("Failed to load keys", exc_info=True)
-        return 1
+    auth, server_keys = prepare_auth(ctx, os.path.expanduser('~/.dync'))
```
Comment (aseyboldt) [inline dync/server.py]: Should we move them to `/etc/dync`? (Not related to this pull request)

File: dync/daemon.py [hunk 8]
```diff
@@ -96,17 +95,20 @@ def stop(self):
 
         # Try killing the daemon process
         try:
-            while 1:
-                os.kill(pid, signal.SIGTERM)
-                time.sleep(0.1)
+            os.kill(pid, 0)
```
Comment (aseyboldt) [inline dync/daemon.py]: No need to check if the process exists *before* you send sigterm. (btw: in line 90 it should be `if pid is None`, not a real problem in this case as the pid should never be 0)

File: dync/daemon.py [hunk 9]
```diff
@@ -96,17 +95,20 @@ def stop(self):
...
             if e.find("No such process") > 0:
                 if os.path.exists(self._pidfile):
                     os.remove(self._pidfile)
             else:
-                print(str(err.args))
+                sys.stderr.write(str(err.args) + "\n")
                 sys.exit(1)
+        time.sleep(0.1)
+        # We checked if the process is present, so kill it
+        os.kill(pid, signal.SIGTERM)
```
Comment (aseyboldt) [inline dync/daemon.py]: see comment.

File: dync/daemon.py (hunk 8 above)
Comment (sven1103) [inline dync/daemon.py]: Done

File: dync/server.py [hunk 10]
```diff
@@ -332,42 +325,64 @@ def init(config):
...
+    args = parser.parse_args()
+
+    return args
 
 
 def main():
 
+    try:
+        args = parse_args()
+    except Exception as exc:
```
Comment (sven1103) [inline dync/server.py]: \*hust\*

**Comment by aseyboldt** [comment]
The code for killing the process is still not really what we want. Something like this (not tested):
```
def check_process(pid):
    try:
        os.kill(pid, 0)
    except ProcessLookupError:
        return False
    else:
        return True

def stop(self):
    ....  # read pidfile and return if it does not exist (and catch PermissionError for open!)

    try:
        os.kill(pid, signal.SIGTERM)
    except ProcessLookupError:
        try:
            self.delpidfile()
        except FileNotFoundError:
            pass
        except PermissionError:
            print("Could not remove pidfile. Permission denied.")
        return

    time = 0
    while time < 10:
        time.sleep(0.1)
        time += 0.1
        if not check_process(pid):
            break
    else:
        print("Deamon does not shutdown on SIGTERM, killing it...", file=sys.stderr)
        os.kill(pid, signal.SIGKILL)

    try:
        self.delpidfile()
    except FileNotFoundError:
        pass
    except PermissionError:
        print("Could not remove pidfile. Permission denied.")
```

and wrap the `daemon.stop()` in an `try: ... except PermissionError:`

File: dync/server.py (hunk 7 above)
Comment (sven1103) [inline dync/server.py]: Continued on #8

**Comment by sven1103** [comment]
ah, now i get what you meant. You give the sigterm some time to shutdown, and after some cadence, you throw an Exception because the process most likely is not going to be killed :) Say that instantly next time :P

**Comment by aseyboldt** [comment]
Well, not throw an exception, but kill it with sigkill (which is just saying to the os: "Don't tell this process anything, just don't give it any time on the cpu anymore and mark it as finished.")

File: dync/daemon.py (hunk 9 above)
Comment (sven1103) [inline dync/daemon.py]: Done

File: dync/server.py [hunk 11]
```diff
@@ -370,15 +370,23 @@ def main():
                 log.exception("Starting failed.")
                 sys.exit(1)
         elif args.command == "stop":
-            dync_dameon.stop()
+            try:
+                dync_dameon.stop()
+            except PermissionError:
+                log.error("Permisson denied for pidfile")
```
Comment (aseyboldt) [inline dync/server.py]: Could also come from os.kill.

File: dync/server.py [hunk 12]
```diff
@@ -370,15 +370,23 @@ def main():
...
+                sys.exit(1)
         elif args.command == "restart":
             try:
                 dync_dameon.restart(init, config)
+            except PermissionError:
+                log.error("Permisson denied for pidfile")
```
Comment (aseyboldt) [inline dync/server.py]: same

File: service/dync.sh [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+#
+# Init script for dyncserver
+#
+# chkconfig:   - 30 70
+# description: Redirects incoming files to target openBis dropbox
+
+# Source function library.
+. /etc/rc.d/init.d/functions
+
+exec="/usr/bin/dyncserver"
```
Comment (aseyboldt) [inline service/dync.sh]: did you rename the entry point? I think it used to be dync-server

File: service/dync.sh [hunk 14]
```diff
@@ -0,0 +1,117 @@
+#!/bin/sh
```
Comment (aseyboldt) [inline service/dync.sh]: No `.sh` in filename

File: service/dync.sh (hunk 13 above)
Comment (sven1103) [inline service/dync.sh]: sry, typo

**Comment by aseyboldt** [comment]
I think we can merge this now. Probably still needs some testing, but I think we can deal with problems than may come up in new issues/pull requests.

**Comment by sven1103** [comment]
Wuhuuu

**Pull request merged**